From 918703db89c6c6b5778670dd5b8c28b44bde9046 Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@st.com>
Date: Wed, 10 Apr 2024 11:07:43 +0200
Subject: [PATCH 570/841] feat(st): add a function to clear the FWU trial state
 counter

Add an API stm32_clear_fwu_trial_boot_cnt() function to clear the
trial state counter. This is called in the corner case scenario when
the active index is in an Invalid state, thus needing a reset of the
trial state counter.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git dbf4860f099a708f3cb05c75cbbda561eeb99b70 feat(st): add a function to clear the FWU trial state counter]
Change-Id: I2980135da88d0d947c222655c7958b51eb572d69
Signed-off-by: Sughosh Ganu <sughosh.ganu@linaro.org>
(cherry picked from commit 6e99fee43efa256bdac3b38864206c94bd9ae3c8)
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/372651
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Tested-by: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 plat/st/common/include/stm32mp_common.h | 1 +
 plat/st/common/stm32mp_common.c         | 8 ++++++++
 2 files changed, 9 insertions(+)

diff --git a/plat/st/common/include/stm32mp_common.h b/plat/st/common/include/stm32mp_common.h
index 71015b565..2c983ada9 100644
--- a/plat/st/common/include/stm32mp_common.h
+++ b/plat/st/common/include/stm32mp_common.h
@@ -191,6 +191,7 @@ int stm32_get_stop2_entrypoint_cell(struct nvmem_cell *stop2_entrypoint);
 void stm32_fwu_set_boot_idx(void);
 uint32_t stm32_get_and_dec_fwu_trial_boot_cnt(void);
 void stm32_set_max_fwu_trial_boot_cnt(void);
+void stm32_clear_fwu_trial_boot_cnt(void);
 #endif /* PSA_FWU_SUPPORT */
 
 #endif /* STM32MP_COMMON_H */
diff --git a/plat/st/common/stm32mp_common.c b/plat/st/common/stm32mp_common.c
index 0f8283ef6..a341cf40e 100644
--- a/plat/st/common/stm32mp_common.c
+++ b/plat/st/common/stm32mp_common.c
@@ -656,5 +656,13 @@ void stm32_set_max_fwu_trial_boot_cnt(void)
 				(FWU_MAX_TRIAL_REBOOT << FWU_INFO_CNT_OFF) &
 					FWU_INFO_CNT_MSK);
 }
+
+void stm32_clear_fwu_trial_boot_cnt(void)
+{
+	struct nvmem_cell fwu_info_cell = {};
+
+	stm32_get_fwu_info_cell(&fwu_info_cell);
+	stm32_nvmem_cell_clrset(&fwu_info_cell, FWU_INFO_CNT_MSK, 0U);
+}
 #endif /* PSA_FWU_SUPPORT */
 #endif /* !STM32MP_SSP */
-- 
2.47.2

